# Exploring eBay Car Sales Data

This project explores a dataset of used car listings scraped from **eBay Kleinanzeigen**, the classifieds section of the German eBay website. The dataset contains **50,000 listings** and includes details such as vehicle brand, model, price, mileage, fuel type, and registration year.

The primary goals of this project are to **clean the raw dataset** — dealing with messy string formatting, missing values, and outliers — and then perform **exploratory analysis** to uncover patterns in the used car market. For example, we'll investigate which brands and models are most common, how price relates to mileage and age, and whether unrepaired damage significantly affects listing prices.

The dataset contains 20 columns covering both vehicle attributes (e.g. `brand`, `model`, `powerPS`, `odometer`) and listing metadata (e.g. `dateCrawled`, `dateCreated`, `lastSeenOnline`). Some columns require cleaning before analysis, as numeric values like `price` and `odometer` are stored as formatted strings.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding='latin1')

In [2]:
autos

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,odometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfl...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,2016-03-27 14:38:19,Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon,privat,Angebot,"$24,900",control,limousine,2011,automatik,239,q5,"100,000km",1,diesel,audi,nein,2016-03-27 00:00:00,0,82131,2016-04-01 13:47:40
49996,2016-03-28 10:50:25,Opel_Astra_F_Cabrio_Bertone_Edition___T�V_ne...,privat,Angebot,"$1,980",control,cabrio,1996,manuell,75,astra,"150,000km",5,benzin,opel,nein,2016-03-28 00:00:00,0,44807,2016-04-02 14:18:02
49997,2016-04-02 14:44:48,Fiat_500_C_1.2_Dualogic_Lounge,privat,Angebot,"$13,200",test,cabrio,2014,automatik,69,500,"5,000km",11,benzin,fiat,nein,2016-04-02 00:00:00,0,73430,2016-04-04 11:47:27
49998,2016-03-08 19:25:42,Audi_A3_2.0_TDI_Sportback_Ambition,privat,Angebot,"$22,900",control,kombi,2013,manuell,150,a3,"40,000km",11,diesel,audi,nein,2016-03-08 00:00:00,0,35683,2016-04-05 16:45:07


In [3]:
autos.info()
autos.head()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   dateCrawled          50000 non-null  str  
 1   name                 50000 non-null  str  
 2   seller               50000 non-null  str  
 3   offerType            50000 non-null  str  
 4   price                50000 non-null  str  
 5   abtest               50000 non-null  str  
 6   vehicleType          44905 non-null  str  
 7   yearOfRegistration   50000 non-null  int64
 8   gearbox              47320 non-null  str  
 9   powerPS              50000 non-null  int64
 10  model                47242 non-null  str  
 11  odometer             50000 non-null  str  
 12  monthOfRegistration  50000 non-null  int64
 13  fuelType             45518 non-null  str  
 14  brand                50000 non-null  str  
 15  notRepairedDamage    40171 non-null  str  
 16  dateCreated          50000 non-nu

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,odometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfl...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


## Initial Observations

From the output of `info()` and `head()`, we can note the following:

- The dataset has **50,000 rows and 20 columns**. Most columns are stored as strings (`str`), including columns that should be numeric (`price`, `odometer`) or datetime (`dateCrawled`, `dateCreated`, `lastSeen`).
- **No column exceeds ~20% null values.** The most incomplete column is `notRepairedDamage` with ~9,800 nulls (~20%), followed by `vehicleType` (~5,100 nulls) and `gearbox` (~2,700 nulls).
- **Column names use camelCase** (e.g. `yearOfRegistration`, `notRepairedDamage`) rather than Python's preferred `snake_case`, and some names are not very descriptive.
- `nrOfPictures` appears to be all zeros — likely a scraping artifact and probably not useful for analysis.

## Cleaning Column Names

Before exploring the data further, we'll rename the columns to follow Python's `snake_case` convention and make a few names more descriptive:

| Original | Renamed |
|---|---|
| `yearOfRegistration` | `registration_year` |
| `monthOfRegistration` | `registration_month` |
| `notRepairedDamage` | `unrepaired_damage` |
| `dateCreated` | `ad_created` |
| All others | converted to `snake_case` |

Using consistent, readable column names makes the code easier to write and understand throughout the rest of the analysis.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='str')

In [5]:
mapping_dict = {
    'dateCrawled': 'date_crawled',
    'name': 'name',
    'seller': 'seller',
    'offerType': 'offer_type',
    'price': 'price',
    'abtest': 'abtest',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'gearbox': 'gearbox',
    'powerPS': 'power_ps',
    'model': 'model',
    'odometer': 'odometer',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'brand': 'brand',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'n_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen'
}

autos.columns = autos.columns.map(mapping_dict)

In [6]:
autos.head()

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,n_of_pictures,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfl...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


## Column Renaming Summary

We renamed all 20 columns using a mapping dictionary and `autos.columns.map()`. Two types of changes were made:

**1. camelCase → snake_case** (Python convention): Column names like `dateCrawled`, `offerType`, and `fuelType` were converted to `date_crawled`, `offer_type`, and `fuel_type`. This makes them consistent with Python's style guide (PEP 8) and easier to type without accidentally triggering case-sensitive bugs.

**2. Descriptive renames**: Four columns were also given clearer names:
- `yearOfRegistration` → `registration_year` — groups both registration columns under a common `registration_` prefix
- `monthOfRegistration` → `registration_month` — same rationale
- `notRepairedDamage` → `unrepaired_damage` — more natural phrasing; the double negative in the original name was easy to misread
- `dateCreated` → `ad_created` — disambiguates it from other date columns; makes clear this is when the *listing* was created, not the car

In [7]:
print(autos.describe(include='all'))
print(autos.info())

               date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
cou

## Descriptive Statistics — Observations

**Columns that are candidates to be dropped:**
- `n_of_pictures` — mean, min, and max are all `0`. Every single listing has zero pictures, making this column completely uninformative.
- `seller` — 49,999 out of 50,000 listings are `privat` (private). Only 1 is a dealer. No analytical value.
- `offer_type` — 49,999 out of 50,000 are `Angebot` (offer). Same issue — no variation.

**Columns needing further investigation:**
- `registration_year` — min is `1000` and max is `9999`, which are clearly erroneous. We'll need to filter to a realistic range.
- `power_ps` — max is `17,700` PS and min is `0`. Both extremes are likely data entry errors.
- `registration_month` — min is `0`, which is not a valid month (months run 1–12).
- `price` — top value is `$0` with a frequency of 1,421. Many listings may have been posted with no price. Needs cleaning.
- `odometer` — only 13 unique values, stored as strings with `km` suffix (e.g. `150,000km`). Needs to be converted to numeric.

**Numeric data stored as text:**
- `price` — stored as a string with a `$` prefix and comma separators (e.g. `$5,000`). Needs the `$` and `,` stripped and conversion to `int`.
- `odometer` — stored as a string with a `km` suffix and comma separators (e.g. `150,000km`). Needs the `km` and `,` stripped and conversion to `int`. The column should then be renamed to `odometer_km` to reflect the unit.

In [8]:
print(autos['price'])
print(autos['odometer'])

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: str
0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: str


In [9]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')

autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')


autos['price'] = autos['price'].astype(float)
autos['odometer'] = autos['odometer'].astype(float)

In [10]:
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

In [11]:
autos.head()

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,n_of_pictures,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,5000.0,control,bus,2004,manuell,158,andere,150000.0,3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,8500.0,control,limousine,1997,automatik,286,7er,150000.0,6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,8990.0,test,limousine,2009,manuell,102,golf,70000.0,7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,4350.0,control,kleinwagen,2007,automatik,71,fortwo,70000.0,6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfl...,privat,Angebot,1350.0,test,kombi,2003,manuell,0,focus,150000.0,7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


In [12]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe(include='all'))
print(autos['odometer_km'].value_counts().head())
print('\n')
print(autos['price'].unique().shape)
print(autos['price'].describe(include='all'))
print(autos['price'].value_counts().head())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: count, dtype: int64


(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
price
0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: count, dtype: int64


In [13]:
print(autos[autos['price'] == 0].count())
print('\n')
print(autos['price'].value_counts().sort_index().head(100))
print(autos['price'].value_counts().sort_index().tail(100))

date_crawled          1421
name                  1421
seller                1421
offer_type            1421
price                 1421
abtest                1421
vehicle_type           915
registration_year     1421
gearbox               1088
power_ps              1421
model                 1126
odometer_km           1421
registration_month    1421
fuel_type              974
brand                 1421
unrepaired_damage      699
ad_created            1421
n_of_pictures         1421
postal_code           1421
last_seen             1421
dtype: int64


price
0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
         ... 
300.0     384
310.0       1
320.0      12
325.0       5
329.0       2
Name: count, Length: 100, dtype: int64
price
65699.0       1
65700.0       1
65990.0       1
66500.0       1
66964.0       1
             ..
10000000.0    1
11111111.0    2
12345678.0    3
27322222.0    1
99999999.0    1
Name: count, Length: 100, dtype: int64


## Outlier Removal — `price` and `odometer_km`

**`price`**

Exploring the low end revealed **1,421 listings priced at $0**, along with a long tail of implausibly cheap listings ($1, $2, $3, etc.). These are almost certainly placeholder values or test listings rather than real sales. On the high end, values like `$11,111,111`, `$12,345,678`, and `$99,999,999` are clearly bogus — no used car on a German classifieds site sells for tens of millions of dollars.

We removed all listings outside the range **$500–$350,000**, dropping ~4,900 rows. This brought the dataset from 50,000 to **45,097 rows**.

After filtering, the distribution looks much more realistic:
- **Median price: $3,500** — consistent with an affordable used car market
- **Mean price: $6,321** — pulled slightly higher by luxury and newer vehicles
- **75th percentile: $7,900** — the majority of listings are budget-friendly

**`odometer_km`**

The `odometer_km` column only contains **13 unique values**, all in round-number increments (e.g. `5,000km`, `10,000km`, `20,000km`, up to `150,000km`). This suggests mileage was entered via a dropdown selector rather than free text, so there are no true outliers. The most common value by far is `150,000km` (~32,000 listings), meaning the majority of cars listed are high-mileage. No filtering was needed for this column.

In [14]:
autos = autos[autos['price'].between(500, 350000)]

In [15]:
print(autos['price'].describe())

count     45097.000000
mean       6320.659600
std        9261.841444
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      350000.000000
Name: price, dtype: float64


In [16]:
print(autos['date_crawled'].str[:10])
print(autos['ad_created'].str[:10])
print(autos['last_seen'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 45097, dtype: str
0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 45097, dtype: str
0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 45097, dtype: str


## Date Column Observations

**`date_crawled`, `ad_created`, `last_seen`**

All three crawler/listing date columns span roughly **March 5, 2016 to April 7, 2016** — a window of about 5 weeks. The proportions are fairly evenly distributed across dates, with no single day dominating. The final days (April 5–7) appear slightly higher in `last_seen`, which makes sense as the crawler would have seen recently-active listings more recently. No cleaning is needed for these columns.

**`registration_year`**

The distribution here is deeply problematic. The **minimum is 1000** and the **maximum is 9999** — both are clearly impossible registration years for road vehicles. The mean (~2005) and median (2004) look reasonable, suggesting the bulk of the data is valid, but the extreme outliers need to be removed before this column can be used analytically.

Since the data was scraped in 2016, no car could have a registration year later than 2016. On the lower end, mass-produced cars didn't exist before the early 1900s, so a reasonable lower bound would be around 1900. We should filter `registration_year` to a realistic range before proceeding.

In [17]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))

date_crawled
2016-04-07    0.001353
2016-04-06    0.003171
2016-04-05    0.013172
2016-04-04    0.036610
2016-04-03    0.038827
2016-04-02    0.035767
2016-04-01    0.033905
2016-03-31    0.031665
2016-03-30    0.033328
2016-03-29    0.033262
2016-03-28    0.034836
2016-03-27    0.031177
2016-03-26    0.032641
2016-03-25    0.031089
2016-03-24    0.028982
2016-03-23    0.032397
2016-03-22    0.033018
2016-03-21    0.037741
2016-03-20    0.038073
2016-03-19    0.034747
2016-03-18    0.012883
2016-03-17    0.031155
2016-03-16    0.029359
2016-03-15    0.034016
2016-03-14    0.036300
2016-03-13    0.015522
2016-03-12    0.037320
2016-03-11    0.033018
2016-03-10    0.032707
2016-03-09    0.032907
2016-03-08    0.033173
2016-03-07    0.036189
2016-03-06    0.014125
2016-03-05    0.025567
Name: proportion, dtype: float64
ad_created
2016-04-07    0.001197
2016-04-06    0.003260
2016-04-05    0.011908
2016-04-04    0.036987
2016-04-03    0.039049
                ...   
2015-12-05    0.000022


In [18]:
print(autos['registration_year'].describe())

count    45097.000000
mean      2005.064173
std         89.652017
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


In [19]:
autos[autos['registration_year'] < 1930] # can clearly remove those before 1900


,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,n_of_pictures,postal_code,last_seen
21416,2016-03-12 08:36:21,Essex_super_six__Ford_A,privat,Angebot,16500.0,control,cabrio,1927,manuell,40,andere,5000.0,5,benzin,ford,NaN,2016-03-12 00:00:00,0,74821,2016-03-15 12:45:12
22101,2016-03-09 16:51:17,BMW_Andere,privat,Angebot,11500.0,test,cabrio,1929,manuell,15,andere,5000.0,1,NaN,bmw,ja,2016-03-09 00:00:00,0,70569,2016-04-07 06:17:11
22316,2016-03-29 16:56:41,VW_Kaefer.__Zwei_zum_Preis_von_einem.,privat,Angebot,1500.0,control,NaN,1000,manuell,0,kaefer,5000.0,0,benzin,volkswagen,NaN,2016-03-29 00:00:00,0,48324,2016-03-31 10:15:28
22659,2016-03-14 08:51:18,Opel_Corsa_B,privat,Angebot,500.0,test,NaN,1910,NaN,0,corsa,150000.0,0,NaN,opel,NaN,2016-03-14 00:00:00,0,52393,2016-04-03 07:53:55
28693,2016-03-22 17:48:41,Renault_Twingo,privat,Angebot,599.0,control,kleinwagen,1910,manuell,0,NaN,5000.0,0,benzin,renault,NaN,2016-03-22 00:00:00,0,70376,2016-04-06 09:16:59
49283,2016-03-15 18:38:53,Citroen_HY,privat,Angebot,7750.0,control,NaN,1001,NaN,0,andere,5000.0,0,NaN,citroen,NaN,2016-03-15 00:00:00,0,66706,2016-04-06 18:47:20


## Filtering `registration_year`

**Chosen bounds: 1900–2016**

- **Upper bound (2016):** Since the data was scraped in 2016, no car could legally have a registration year beyond that. There were **1,772 rows** with years after 2016 — a significant and clearly erroneous group that must be removed.
- **Lower bound (1900):** Inspecting values before 1930 revealed only **6 rows**, with years like 1910, 1927, and 1929. While technically possible (vintage cars do exist), these are so rare they're likely data entry errors. Setting the lower bound at 1900 excludes the genuinely impossible values (e.g. year 1000) while retaining any legitimate historic vehicles.

## Distribution of `registration_year` After Filtering

The cleaned distribution shows a clear and expected pattern — listings are heavily concentrated in the **1995–2010** range, peaking around **1999–2005**. Pre-1980 vehicles make up a negligible fraction (under 0.1% combined), confirming they were outliers. The data now represents a realistic snapshot of the used car market: mostly vehicles that are 5–20 years old at the time of listing.

In [20]:
autos[autos['registration_year'] > 2016].count() # registration year well after the date of scrapping is concerning, even more so by the vast number of them

date_crawled          1772
name                  1772
seller                1772
offer_type            1772
price                 1772
abtest                1772
vehicle_type             2
registration_year     1772
gearbox               1563
power_ps              1772
model                 1525
odometer_km           1772
registration_month    1772
fuel_type             1104
brand                 1772
unrepaired_damage     1024
ad_created            1772
n_of_pictures         1772
postal_code           1772
last_seen             1772
dtype: int64

In [21]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [22]:
autos['registration_year'].value_counts(normalize=True).sort_index()

registration_year
1910    0.000046
1927    0.000023
1929    0.000023
1931    0.000023
1934    0.000046
          ...   
2012    0.030169
2013    0.018397
2014    0.015050
2015    0.008402
2016    0.021928
Name: proportion, Length: 78, dtype: float64

In [23]:
brand_counts = autos['brand'].value_counts(normalize=True)
top_brands = brand_counts[brand_counts >= 0.05].index

In [24]:
brand_dict = {}

for b in top_brands:
    brand_dict[b] = autos[autos['brand'] == b]['price'].mean()
        

pd.Series(brand_dict).sort_values(ascending=False)

audi             9613.647794
mercedes_benz    8766.902709
bmw              8582.261288
volkswagen       5783.622985
ford             4291.666312
opel             3394.039568
dtype: float64

In [25]:
brand_table = pd.Series(brand_dict).to_frame('mean_price')

## Brand Analysis — Mean Price and Listing Count

We filtered to the 6 brands that each represent more than 1% of total listings, covering the dominant players in the German used car market.

**Volume:** Volkswagen dominates with 9,180 listings — more than double the next highest brand (BMW at 4,983). This aligns with VW's status as Germany's most popular domestic brand. Audi and Ford have the fewest listings of the group at ~3,900 and ~2,800 respectively.

**Price:** There's a clear split between budget and premium brands:
- **Audi** commands the highest mean price at **~€9,614**, followed closely by **Mercedes-Benz (~€8,767)** and **BMW (~€8,582)** — all three are premium German manufacturers
- **Volkswagen** sits in the middle at **~€5,784**, reflecting its broad range from entry-level to higher-end models
- **Opel (~€3,394)** and **Ford (~€4,292)** are the most affordable on average, consistent with their positioning as mass-market budget brands

The data suggests a strong correlation between brand prestige and average listing price, even in the used car market.

In [26]:
brand_table

,mean_price
volkswagen,5783.622985
bmw,8582.261288
mercedes_benz,8766.902709
opel,3394.039568
audi,9613.647794
ford,4291.666312


## Mean Price vs. Mean Mileage by Brand

Adding average mileage alongside mean price reveals an interesting pattern — mileage is remarkably consistent across all six brands, ranging from ~123,000km (Ford) to ~133,000km (BMW), a spread of only ~10,000km. Despite this, there are dramatic differences in mean price.

This tells us that **mileage alone does not explain the price gap between brands**. BMW and Mercedes-Benz listings actually have the *highest* average mileage yet still command prices more than double those of Opel and Ford. The price differences are better explained by **brand prestige and perceived quality** rather than how worn-out the cars are.

Opel stands out as the cheapest brand on average (~€3,394) despite having similar mileage to the premium brands — suggesting buyers heavily discount Opel regardless of condition. Audi leads on price (~€9,614) with middle-of-the-pack mileage, reinforcing its premium positioning in the used car market.

In [27]:
brand_table['brand_count'] = autos['brand'].value_counts()

In [28]:
mileage_dict = {}


for b in top_brands:
    mileage_dict[b] = autos[autos['brand'] == b]['odometer_km'].mean()

brand_table['avg_mileage'] = mileage_dict

In [29]:
brand_table

,mean_price,brand_count,avg_mileage
volkswagen,5783.622985,9180,128234.749455
bmw,8582.261288,4983,132865.743528
mercedes_benz,8766.902709,4430,131019.187359
opel,3394.039568,4347,128012.422360
audi,9613.647794,3921,128909.716909
ford,4291.666312,2823,123494.509387
